### ISSUE TITLE
Do not allow changing parameter units in model instances

### ISSUE THREAD

**Issue by nden** [astropy/astropy#7224]
Raised by @perrygreenfield in #7204 

One thing I think should change is that it should not be possible to assign new units to a Parameter instance (E.g. this currently works:

g = Gaussian1D(1 * u.J, 1 * u.m, 0.1 * u.m)

g.amplitude = 2 * u.s
g.mean = 2 * u.Jy

I can't see the justification for allowing this. If a model with different unit support is needed, make a new one, don't mutate an existing one. Seems like this is asking for trouble.

**Comment by nden** [comment]
I agree this should not be allowed in in general with arbitrary units. What about the case when the new unit can be converted to the original one? It seems like a useful option.

**Comment by perrygreenfield** [comment]
I'd agree if the assignment behavior converts the value to the original unit. For example, in the above case I do g.mean = 1. * u.cm, then what happens is that it is converted to u.m, and the stored value is 100, and the unit is not changed. I say this because it is possible that the model definition is just using the value assuming the actual unit doesn't change. Would this behavior be acceptable?

**Comment by nden** [comment]
@perrygreenfield I think that's reasonable.

@astrofrog @Cadair Any thoughts on this?

**Comment by perrygreenfield** [comment]
Um, I meant the value that is stored is 0.01 u.m. I can't do unit conversions anymore and must rely on astropy units...

**Comment by astrofrog** [comment]
I'm a little worried that we are going to head back in a direction we moved away from in terms of input validation of the units. We decided a little while back to just leave it up to the evaluate method to figure out if the units are ok. In this case, what if I have a Gaussian with a mean of ``3 * u.m`` and I want to change it to ``3 * u.Hz`` - would this be allowed? How would we indicate which equivalencies are ok? To flip the question, what is the harm in allowing the units to be changed? Inconsistencies will be picked up at evaluation time anyway.

**Comment by perrygreenfield** [comment]
@astrofrog Inconsistencies won't be picked up in the following cases:

- The model is not intrinsically using Quantities but just the value assumed to be a particular unit.
- In cases where the parameter unit is arbitrary (i.e., does not affect the computation at all; a good example is amplitude in Gaussian1D) it won't result in an error at evaluation. In cases where the output unit isn't critical that isn't a problem. But supposing the model is promising a physical quantity, it would simply be wrong.

Handling equivalencies could be done for parameters like anything else by assigning an attribute listing allowed equivalencies. 

I figure is someone defined a model with specific units, they had a good reason to do so and it shouldn't be permitted to fiddle with them.
